### ISSUE TITLE
Back propagation behaviour

### ISSUE THREAD

**Issue by BDUG** [huggingface/candle#1509]
I tried almost all given example such as the following https://github.com/huggingface/candle/blob/main/candle-examples/examples/mnist-training/main.rs 

In all example I see a chain of calculations representing the forward step e.g., 

let xs = xs
            .reshape((b_sz, 1, 28, 28))?
            .apply(&self.conv1)?
            .max_pool2d(2)?
            .apply(&self.conv2)?
            .max_pool2d(2)?
            .flatten_from(1)?
            .apply(&self.fc1)?
            .relu()?;

But I do see only one back step, which seem to address only the last "layer" of the DNN e.g.,

let logits = model.forward(&train_images, true)?;
let log_sm = ops::log_softmax(&logits, D::Minus1)?;
let loss = loss::nll(&log_sm, &train_labels)?;
opt.backward_step(&loss)?;

Does this the usual back propagation or do I totally got something wrong?

**Comment by s-casci** [comment]
From `training_loop_cnn` in `mnist-training/main.rs`:
```rust
let mut varmap = VarMap::new();
let vs = VarBuilder::from_varmap(&varmap, DType::F32, &dev);
let model = ConvNet::new(vs.clone())?;
```
As you can see the model is passed a `VarBuilder` when constructed, which is then used to store the parameters (like weights and biases) of its layers.

For example, the first convolutional layer of the `ConvNet` is stored like:
```rust
let conv1 = candle_nn::conv2d(1, 32, 5, Default::default(), vs.pp("c1"))?;
```
where `vs` is the `VarBuilder` variable passed in the constructor.

Then, the optimizer, in order to compute and propagate the loss for all layers (not just the last one), is passed the parameters of the network, like so:
```rust
let mut opt = candle_nn::AdamW::new(varmap.all_vars(), adamw_params)?;
```
since `varmap.all_vars()` returns them.

If you are familiar with `PyTorch`, the effect is similar to calling the `get_parameters` method of a `torch.nn.Module`.

I hope this clarifies how backpropagation works in Candle. I'd suggest to take a look at the `VarMap` and `VarBuilder` implementations for further details.

**Comment by EricLBuehler** [comment]
Great answer by s-casci, I just wanted to elaborate on the backprop step:

Throughout the forward step, operations on variable `Tensor`s are tracked. Because your models operates on `Tensors`s (that are variables), all operations that you perform in the forward step will be converted into a graph (a DAG, specifically). This is used to calculate the gradient. The optimizer simply takes those gradient values and applies them to the tensor.

**Comment by BDUG** [comment]
@s-casci  @EricLBuehler many thanks for your great answers. Got it.
